solidity/contracts/contracts/QuantumGravityBridge.sol

```diff
@@ -0,0 +1,314 @@
+            if (!verifySig(_currentValidators[i].addr, _digest, _sigs[i])) {
+                revert InvalidSignature();
+            }
+
+            // Sum up cumulative power.
+            cumulativePower += _currentValidators[i].power;
+
+            // Break early to avoid wasting gas.
+            if (cumulativePower >= _powerThreshold) {
+                break;
+            }
+        }
+
+        // Check that there was enough power
+        if (cumulativePower < _powerThreshold) {
+            revert InsufficientVotingPower();
+        }
+    }
+
+    /// @notice This updates the validator set by checking that the validators
+    /// in the current validator set have signed off on the new validator set.
+    /// The signatures supplied are the signatures of the current validator set
+    /// over the checkpoint hash generated from the new validator set. Anyone
+    /// can call this function, but they must supply valid signatures of the
+    /// current validator set over the new validator set.
+    /// @param _newValidatorSetHash The hash of the new validator set
+    /// @param _newNonce The new nonce.
+    /// @param _currentValidatorSet The current validator set.
+    /// @param _sigs Signatures.
+    function updateValidatorSet(
+        uint256 _newNonce,
+        uint256 _newPowerThreshold,
+        bytes32 _newValidatorSetHash,
+        Validator[] calldata _currentValidatorSet,
+        Signature[] calldata _sigs
+    ) external {
+        // CHECKS
+
+        uint256 currentNonce = s_lastValidatorSetNonce;
+        uint256 currentPowerThreshold = s_powerThreshold;
+
+        // Check that the valset nonce is greater than the old one.
+        if (_newNonce <= currentNonce) {
+            revert InvalidValidatorSetNonce();
+        }
+
+        // Check that current validators and signatures are well-formed.
+        if (_currentValidatorSet.length != _sigs.length) {
+            revert MalformedCurrentValidatorSet();
+        }
+
+        // Check that the supplied current validator set matches the saved checkpoint
+        bytes32 currentValsetHash = keccak256(abi.encode(_currentValidatorSet));
+        if (
+            domainSeparateValidatorSetHash(BRIDGE_ID, currentNonce, currentPowerThreshold, currentValsetHash) !=
+            s_lastValidatorSetCheckpoint
+        ) {
+            revert SuppliedValidatorSetInvalid();
+        }
+
+        // Check that enough current validators have signed off on the new validator set
+        bytes32 newCheckpoint = domainSeparateValidatorSetHash(
+            BRIDGE_ID,
+            currentNonce,
+            _newPowerThreshold,
+            _newValidatorSetHash
+        );
+        checkValidatorSignatures(_currentValidatorSet, _sigs, newCheckpoint, currentPowerThreshold);
+
+        // EFFECTS
+
+        s_lastValidatorSetCheckpoint = newCheckpoint;
+        s_powerThreshold = _newPowerThreshold;
+        s_lastValidatorSetNonce = _newNonce;
```

Comment:
This is quite different from https://github.com/celestiaorg/quantum-gravity-bridge/blob/b5ba246f23f89c43adf764e2e3f804b4a9f3beb0/solidity/contracts/Peggy.sol#L234-L318

Here we only take a hash of the new valSet while in the original contract we actually took-ib the new validator set 🤔 Is that what you meant with optimistic?

Reply:
No, this is one of the simple low-hanging gas optimizations. Under an honest majority assumption, you only need to check the "checkpoint" was signed over, not its contents.